WIP: flux-mini: add environment manipulation options

grondo opened the pull request:
Ok, I'm parking this experimental PR here before moving on to other things.

This PR adds `--env=RULE`, `--env-filter=PATTERN`, and `--env-file=FILE` options to `flux-mini` as discussed in #3141. 

The PR is marked WIP for now because there were some concerns about the interface. Namely this interface doesn't allow exporting environment variables that begin with `-` or `^` explictly to jobs, but you can't do that with the shell anyway (at least most shells). As @trws demonstrated a user could run `flux-mini` under `/usr/bin/env` if an environment variable not already set in the environment needs to start with a `-` or `^`.

That being said, this interface was just an experiment so I'd be fine trying something else.

Note that env variable values are expanded before being set by `string.Template` using a `ChainMap` of the current jobspec environment and `os.environ`, so this works as you'd expect:

```
$ flux mini run --dry-run --env-filter=* --env='PATH=/foo:${PATH}' --env='PATH=${PATH}:/foo' hostname | jq .attributes.system.environment
{
  "PATH": "/foo:/home/grondo/git/flux-core.git/src/cmd:/usr/bin:/bin:/foo"
}
```

Unfortunately, this requires that the variable expressions be quoted from the shell or else you only get the current environment expanded. Maybe that is too subtle for users?

It is nice when using an env file though, as it allows this to work:
```
-*
FOO=bar
BAR=${FOO}/baz
```

grondo commented:
> One question I couldn't parse out in a quick glance. Is there a way to express a negative rather than positive pattern? Like, I want to remove all variables matching pattern?

Sorry for not being clear @trws.

The current functionality `--env-filter=PATTERN` or `--env=-PATTERN` removes all variables matching a pattern. However, it occurs to me we could treat the normal `VAR` usage, e.g. `--env=VAR` with no "=" as a pattern rule instead of just an exact match.

So this would work, without affecting previous functionality:
```console
$ flux mini run --dry-run --env-filter=* --env='FLUX_*_PATH' hostname | jq .attributes.system.environment
{
  "FLUX_CONNECTOR_PATH": "/home/grondo/git/flux-core.git/src/connectors",
  "FLUX_MODULE_PATH": "/home/grondo/git/flux-core.git/src/modules",
  "FLUX_EXEC_PATH": "/home/grondo/git/flux-core.git/src/broker:/home/grondo/git/flux-core.git/src/cmd",
  "FLUX_PMI_LIBRARY_PATH": "/home/grondo/git/flux-core.git/src/common/.libs/libpmi.so"
}


```

trws commented:
Ah, ok, sorry I had it completely backwards (this is what happens when I try to do too many things at once 🤦 ).  So we have negative patterns, but not positive ones.  Is that true for the regular expression form as well?  I need to think about this a bit, but "filter" in most places I know of usually takes a pattern or lambda or whatever that when true keeps the item, so it might be worth considering another name, or having both positive and negative filters like you mentioned originally?  The C++ std lib is the one place that does the negative pattern thing, they call it `remove_if` in case that sparks any ideas.

trws commented:
Sigh, and I didn’t read this even as well as I should, I like that 
idea of how to do the positive pattern, would we then be able to have 
the regex version be `—env=/<exp>/` for positive rather than 
`—env=-/` for the negative?

On 20 Aug 2020, at 11:19, Mark Grondona wrote:

> > One question I couldn't parse out in a quick glance. Is there a way 
> to express a negative rather than positive pattern? Like, I want to 
> remove all variables matching pattern?
>
> Sorry for not being clear @trws.
>
> The current functionality `--env-filter=PATTERN` or `--env=-PATTERN` 
> removes all variables matching a pattern. However, it occurs to me we 
> could treat the normal `VAR` usage, e.g. `--env=VAR` with no "=" as a 
> pattern rule instead of just an exact match.
>
> So this would work, without affecting previous functionality:
> ```console
> $ flux mini run --dry-run --env-filter=* --env='FLUX_*_PATH' hostname 
> | jq .attributes.system.environment
> {
>   "FLUX_CONNECTOR_PATH": 
> "/home/grondo/git/flux-core.git/src/connectors",
>   "FLUX_MODULE_PATH": "/home/grondo/git/flux-core.git/src/modules",
>   "FLUX_EXEC_PATH": 
> "/home/grondo/git/flux-core.git/src/broker:/home/grondo/git/flux-core.git/src/cmd",
>   "FLUX_PMI_LIBRARY_PATH": 
> "/home/grondo/git/flux-core.git/src/common/.libs/libpmi.so"
> }
>
>
> ```
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/3150#issuecomment-677822564

grondo commented:
Thanks! I was having the same thought about the name! I'm fine with `--env-remove=PATTERN` if you are!
If you have other ideas let me know.

Note that in my current proposal `--env=PATTERN` would be the positive match, where `-PATTERN` or `--env-remove=PATTERN` would be the negative match.

A subtlety though,  `--env=PATTERN` doesn't *only* include environment variables that match a pattern. It would pull all matching variables from the current environment, so it only makes sense when starting with an empty environment. That is, `--env=*` could be considered an implied first '--env' argument for `flux mini`.

trws commented:
> A subtlety though, `--env=PATTERN` doesn't _only_ include environment variables that match a pattern. It would pull all matching variables from the current environment, so it only makes sense when starting with an empty environment. That is, `--env=*` could be considered an implied first '--env' argument for `flux mini`.

That's an interesting point.  It kidna gets back to my earlier thought/question, if a user did `--env-remove=* --env=OMP_*` would that remove everything and start with no environment, or remove everything then whitelist the ones starting with OMP_?  The former is maybe easier to describe, but the latter is what things like `.gitignore` do, and might be quite useful.  It would probably mean it would have to process as: "filter into or out of effective environment from original environment" rather than from current state of effective environment... Not sure.

I think the `env-remove` naming makes it much more clear by the way, works for me!

grondo commented:
> would we then be able to have
the regex version be `—env=/<exp>/` for positive rather than
`—env=-/` for the negative?

Yes, though I was thinking of keeping both, unless you don't think we need a negative match?

grondo commented:
> unless you don't think we need a negative match?

Or, rather, I guess it is more technically "remove all matching entries" (I'm getting confused on my terminology)

trws commented:
Yeah I was thinking have both, shouldn’t have said it as “rather 
than” more meant “as opposed to” or “in addition to” or 
something.

On 20 Aug 2020, at 11:37, Mark Grondona wrote:

> > would we then be able to have
> the regex version be `—env=/<exp>/` for positive rather than
> `—env=-/` for the negative?
>
> Yes, though I was thinking of keeping both, unless you don't think we 
> need a negative match?
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/3150#issuecomment-677830646

grondo commented:
> Yeah I was thinking have both, shouldn’t have said it as “rather
than” more meant “as opposed to” or “in addition to” or
something.

Ok. Let me push that small change here so people could play with it if they want. I'll also change `--env-filter` to `--env-remove`.

grondo commented:
> That's an interesting point. It kidna gets back to my earlier thought/question, if a user did --env-remove=* --env=OMP_* would that remove everything and start with no environment, or remove everything then whitelist the ones starting with OMP_? 

In the current implementation it would be the latter. Both the "positive match" and the template expansion of `$VAR` use a `ChainMap(env, os.environ)`, so lookups are from the built environment first, then fall back to the process environment. The "rules" are stored in a list and applied in the order they are found on the cmdline, with the first implied rule being `*` (so we start with a copy of the current environment).

> The former is maybe easier to describe, but the latter is what things like .gitignore do, and might be quite useful. It would probably mean it would have to process as: "filter into or out of effective environment from original environment" rather than from current state of effective environment... Not sure.

If I understand you correctly, this is pretty much what is done now. I think it successfully follows the principle of least surprise, except where it might not be obvious that we always start with a full environment. (so `--env=FOO` alone actually does nothing -- maybe we need an `--env-only=PATTERN` option that is effectively the same as `--env-remove=* --env=PATTERN`?)

trws commented:
I think this is fine, doing it that way is pretty expressive, and for me 
it’s what I’m used to since the env command uses `-i ONLY_VAR_1=meh` 
to clear and set.  I suppose we could have a shortcut for “clear 
all” like that, but I’m not sure it’s really necessary.

On 20 Aug 2020, at 12:29, Mark Grondona wrote:

> > That's an interesting point. It kidna gets back to my earlier 
> thought/question, if a user did --env-remove=* --env=OMP_* would that 
> remove everything and start with no environment, or remove everything 
> then whitelist the ones starting with OMP_?
>
> In the current implementation it would be the latter. Both the 
> "positive match" and the template expansion of `$VAR` use a 
> `ChainMap(env, os.environ)`, so lookups are from the built environment 
> first, then fall back to the process environment. The "rules" are 
> stored in a list and applied in the order they are found on the 
> cmdline, with the first implied rule being `*` (so we start with a 
> copy of the current environment).
>
>> The former is maybe easier to describe, but the latter is what things 
>> like .gitignore do, and might be quite useful. It would probably mean 
>> it would have to process as: "filter into or out of effective 
>> environment from original environment" rather than from current state 
>> of effective environment... Not sure.
>
> If I understand you correctly, this is pretty much what is done now. I 
> think it successfully follows the principle of least surprise, except 
> where it might not be obvious that we always start with a full 
> environment. (so `--env=FOO` alone actually does nothing -- maybe we 
> need an `--env-only=PATTERN` option that is effectively the same as 
> `--env-remove=* --env=PATTERN`?)
>
> -- 
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/flux-framework/flux-core/pull/3150#issuecomment-677856929

grondo commented:
Great, thanks for the valuable feedback @trws!

SteVwonder commented on src/cmd/flux-mini.py:
```diff
@@ -71,6 +184,33 @@ def create_parser(exclude_io=False):
+        parser.add_argument(
+            "--env-file",
+            action=EnvFileAction,
+            help="Read a set of environment rules from FILE.",
```
```suggestion
            help="Read a set of environment rules from FILE. (multiple use OK)",
```

SteVwonder commented on src/cmd/flux-mini.py:
```diff
@@ -71,6 +184,33 @@ def create_parser(exclude_io=False):
+            + "(see --env-file). Otherwise, set matching environment variables "
+            + "from the current environment (--env=PATTERN) or set a value "
+            + "explicitly (--env=VAR=VALUE). Rules are applied in the order "
+            + "they are used on the command line.",
```
```suggestion
            + "they are used on the command line. (multiple use OK)",
```

SteVwonder commented on src/cmd/flux-mini.py:
```diff
@@ -23,6 +27,115 @@
+    if pattern.startswith("/"):
+        pattern = pattern[1::].rstrip("/")
+    else:
+        pattern = fnmatch.translate(pattern)
```
Oooooo.  This is really nice. TIL!

SteVwonder commented on src/cmd/flux-mini.py:
```diff
@@ -23,6 +27,116 @@
+                #
+                #  Template lookup: use jobspec environment first, fallback
+                #   to current process environment using ChainMap:
+                lookup = ChainMap(environ, os.environ)
+                environ[var] = Template(rest[0]).substitute(lookup)
```
😍 🤤 🥳 💯 🎉 this is soooo clean! I love it!

SteVwonder replied:
One minor thing (sorry for not catching this earlier), can we wrap this in a try-catch to produce a slightly nicer error message?

```
❯ flux mini run --dry-run --env-remove="*" --env='FOO=foo;${BAZ:-baz}' hostname | jq .attributes                                                                                                                               14:30:36 ()
flux-mini: ERROR: Invalid placeholder in string: line 1, col 5

herbein1@quartz2498 ~/Repositories/flux-framework/flux-core (issue#3141 ?S)
❯ FLUX_PYCLI_LOGLEVEL=0 flux mini run --dry-run --env-remove="*" --env='FOO=foo;${BAZ:-baz}' hostname | jq .attributes                                                                                                         14:30:41 ()
flux-mini: ERROR: Invalid placeholder in string: line 1, col 5
flux-mini: DEBUG: Traceback (most recent call last):
  File "/g/g0/herbein1/Repositories/flux-framework/flux-core/src/bindings/python/flux/util.py", line 136, in __call__
    main_func()
  File "/g/g0/herbein1/Repositories/flux-framework/flux-core/src/cmd/flux-mini.py", line 696, in main
    args.func(args)
  File "/g/g0/herbein1/Repositories/flux-framework/flux-core/src/cmd/flux-mini.py", line 434, in main
    jobid = self.submit(args)
  File "/g/g0/herbein1/Repositories/flux-framework/flux-core/src/cmd/flux-mini.py", line 279, in submit
    jobspec.environment = get_filtered_environment(args.env)
  File "/g/g0/herbein1/Repositories/flux-framework/flux-core/src/cmd/flux-mini.py", line 99, in get_filtered_environment
    environ[var] = Template(rest[0]).substitute(lookup)
  File "/usr/lib64/python3.6/string.py", line 130, in substitute
    return self.pattern.sub(convert, self.template)
  File "/usr/lib64/python3.6/string.py", line 127, in convert
    self._invalid(mo)
  File "/usr/lib64/python3.6/string.py", line 103, in _invalid
    (lineno, colno))
ValueError: Invalid placeholder in string: line 1, col 5
```

SteVwonder replied:
Maybe something like this?

```diff
diff --git a/src/cmd/flux-mini.py b/src/cmd/flux-mini.py
index dc0d848dd..9e9e296d5 100755
--- a/src/cmd/flux-mini.py
+++ b/src/cmd/flux-mini.py
@@ -96,7 +96,11 @@ def get_filtered_environment(rules, environ=None):
                 #  Template lookup: use jobspec environment first, fallback
                 #   to current process environment using ChainMap:
                 lookup = ChainMap(environ, os.environ)
-                environ[var] = Template(rest[0]).substitute(lookup)
+                try:
+                    environ[var] = Template(rest[0]).substitute(lookup)
+                except ValueError as e:
+                    LOGGER.error("Failure to substitute env var")
+                    raise
     return environ
```

```
❯ flux mini run --dry-run --env-remove="*" --env='FOO=foo;${BAZ:-baz}' hostname | jq .attributes                                                                                                                               14:56:42 ()
flux-mini: ERROR: Failure to substitute env var
flux-mini: ERROR: Invalid placeholder in string: line 1, col 5
```

grondo replied:
Great idea!

grondo replied:
I think there is also an option to Template.substitute() to either substitute an empty string or leave the unexpanded text alone. I think I prefer generating an error as you have above.

grondo replied:
`Template().substitute()` also raises `KeyError` if `$VAR` is not found. I think this should still be an error, so how about:
```diff
diff --git a/src/cmd/flux-mini.py b/src/cmd/flux-mini.py
index e3a3682e6..c84928933 100755
--- a/src/cmd/flux-mini.py
+++ b/src/cmd/flux-mini.py
@@ -96,7 +96,13 @@ def get_filtered_environment(rules, environ=None):
                 #  Template lookup: use jobspec environment first, fallback
                 #   to current process environment using ChainMap:
                 lookup = ChainMap(environ, os.environ)
-                environ[var] = Template(rest[0]).substitute(lookup)
+                try:
+                    environ[var] = Template(rest[0]).substitute(lookup)
+                except ValueError as e:
+                    LOGGER.error(f"--env: Unable to substitute {rule}")
+                    raise
+                except KeyError as e:
+                    raise Exception(f"--env: Variable {e} not found in {rule}")
     return environ
```

```
ƒ(s=4,d=0,builddir) grondo@asp:~$ flux mini run --dry-run --env-remove="*" --env='FOO=foo;${BAR}' hostname
flux-mini: ERROR: --env: Variable 'BAR' not found in FOO=foo;${BAR}
ƒ(s=4,d=0,builddir) grondo@asp:~$ flux mini run --dry-run --env-remove="*" --env='FOO=foo;${BAZ:-baz}' hostname
flux-mini: ERROR: --env: Unable to substitute FOO=foo;${BAZ:-baz}
flux-mini: ERROR: Invalid placeholder in string: line 1, col 5
```

SteVwonder replied:
Love it.  Also, I totally forgot that we can use f-strings now that we are 3.6+.  Thanks for that reminder and subsequent burst of joy and happiness @grondo :)

grondo replied:
Sadly, the version of pylint in travis doesn't like f-string:
```
W1202: Use % formatting in logging functions and pass the % parameters as arguments (logging-format-interpolation)
```

grondo commented:
Thanks @SteVwonder and @trws! I'll apply @SteVwonder's suggestions, then if the interface looks basically Ok, I'll expand tests and then amend `flux-mini(1)` documentation.

grondo commented:
Ok, I've force pushed another version here with some documentation, and the `try` `catch` block around the template substitution as suggested by @SteVwonder (still need to add tests for error handling there)

grondo commented:
I've pushed the changes to throw an exception when variable substitution fails as noted by @SteVwonder (thanks for catching that!!), added tests and an additional `ENVIRONMENT` section the `flux-mini(1)` manpage.

I've removed WIP for now as this is ready for a real review.
